File: cluster/stream/faiss_api.jl
ENV["JULIA_PYTHONCALL_EXE"] = "/home/zhangyong/miniconda3/bin/python"

using PythonCall

sys = pyimport("sys")
np = pyimport("numpy")
sys.path.insert(0, pystr("/home/zhangyong/codes/julia_learn/cluster/stream"))
faiss_api = pyimport("faiss_api")
# println(sys.path)


function test()
    dir_1 = "/mnt/zy_data/data/longhu_1/sorted_2/"
    feats = np.load(joinpath(dir_1, "feats.npy"))
    println(typeof(feats), feats.shape)
    # feats 需要从julia对象转换为py对象
    index = faiss_api.get_index(384, "")   # L2 index
    println(typeof(index))
    index.add(feats)
    dist, idx = index.search(feats[0:10], 5)  # size:(11,384),因为[0:10]是julia对象
    println(1 - dist / 2)

    query = feats[0:10]    # py对象
    query = pyconvert(Array{Float32, 2}, query)   # juia对象
    # query = ones(Float32, (10, 384))  # Array{Float32}()
    query = np.array(pyrowlist(query), dtype=np.float32)   # py对象
    println(query.shape)
    gallery = feats
    feat_dim = query.shape[1]
    index = faiss_api.get_index(feat_dim, "")
    idxs, dists = faiss_api.rank5(index, query, gallery, topk=5)  # py对象
    println(dists)
    dists = pyconvert(Array{Float32, 2}, dists)
    idxs = pyconvert(Array{Int64, 2}, idxs)

end


# test()


#=
问题:
不支持np.array()的超出总数的idx. 报错 
解决办法: query[pyslice(5,8)]   # idx就是python的 [5,8], 5<=x<8.  

=#

File: cluster/stream/faiss_api.py
import os
import numpy as np
import faiss


def get_index(feat_dim=384, gpus='0'):
    # feat数据存储在这里面. 数据量巨大时,容易爆显存
    os.environ['CUDA_VISIBLE_DEVICES'] = gpus
    # print(os.environ['CUDA_VISIBLE_DEVICES'])
    if gpus == '':
        ngpus = 0
    else:
        ngpus = len(os.environ['CUDA_VISIBLE_DEVICES'].split(','))
    if ngpus == 0:
        cpu_index = faiss.IndexFlatL2(feat_dim)
        gpu_index = cpu_index
    elif ngpus == 1:
        flat_config = faiss.GpuIndexFlatConfig()
        flat_config.device = 0  # int(gpus[0])
        res = faiss.StandardGpuResources()
        gpu_index = faiss.GpuIndexFlatL2(res, feat_dim, flat_config)  # use one gpu.  初始化很慢
    else:
        # print('use all gpu')
        cpu_index = faiss.IndexFlatL2(feat_dim)
        gpu_index = faiss.index_cpu_to_all_gpus(cpu_index)  # use all gpus
    index = gpu_index
    return index


def rank5(index, query, gallery, topk=1000, flag=1):
    """
    增量的add(gallery)
    """
    if flag == 1:
        index.add(gallery)  # <8M
    # print("len(index):", len(index))
    topk = topk  # min(topk, gallery.shape[0])
    dists = np.zeros((query.shape[0], topk), "f")  # 很大.  n*k*4
    idxs = np.zeros((query.shape[0], topk), "int32")  # 很大.  n*k*4
    bs = np.ceil(1e7 * 1.0 / topk)  # 1e6 /  topk=100w/1000  1000, 1w
    bs = int(bs)
    qbatch = int(np.ceil(query.shape[0] * 1.0 / bs))
    # print("#$# qbatch:", qbatch)
    for i in range(qbatch):
        # print(i + 1, "/", qbatch)
        dist, idx = index.search(query[i * bs:(i + 1) * bs], topk)  # L2距离
        dists[i * bs:(i + 1) * bs] = dist
        idxs[i * bs:(i + 1) * bs] = idx
    return idxs, 1 - dists / 2  # 转换为cos相似度


def rank2(query, gallery, topk=1000, metric='cos', ngpus=''):
    # ngpus = 0  # faiss.get_num_gpus()
    # print("number of GPUs:", ngpus)
    os.environ['CUDA_VISIBLE_DEVICES'] = ngpus
    if ngpus == '':
        ngpus = 0
    else:
        ngpus = len(os.environ['CUDA_VISIBLE_DEVICES'].split(','))
    feat_dim = query.shape[1]

    if ngpus == 0:
        cpu_index = faiss.IndexFlatL2(feat_dim)
        gpu_index = cpu_index
    elif ngpus == 1:
        flat_config = faiss.GpuIndexFlatConfig()
        flat_config.device = 0
        res = faiss.StandardGpuResources()
        gpu_index = faiss.GpuIndexFlatL2(res, feat_dim, flat_config)  # use one gpu
    else:
        cpu_index = faiss.IndexFlatL2(feat_dim)
        gpu_index = faiss.index_cpu_to_all_gpus(cpu_index)  # use all gpus
    index = gpu_index

    index.add(gallery)  # <8M
    topk = min(topk, gallery.shape[0])
    dists = np.zeros((query.shape[0], topk), "f")  # 很大.  n*k*4  1m->4G
    idxs = np.zeros((query.shape[0], topk), "int32")  # 很大.  n*k*4
    bs = np.ceil(1e7 * 1.0 / topk)  # 1e6 /  topk=100w/1000  1000, 1w
    bs = int(bs)
    qbatch = int(np.ceil(query.shape[0] * 1.0 / bs))

    # print("#$# qbatch:", qbatch)
    for i in range(qbatch):
        # print(i + 1, "/", qbatch)
        dist, idx = index.search(query[i * bs:(i + 1) * bs], topk)  # L2距离
        dists[i * bs:(i + 1) * bs] = dist
        idxs[i * bs:(i + 1) * bs] = idx

    if metric == 'cos':
        dists = 1 - dists / 2   # 转换为cos相似度
    else:
        dists = dists  # l2 相似度
    return idxs, dists
